system/lib/emmalloc.c

```diff
@@ -865,25 +863,23 @@ void emmalloc_free(void *ptr)
     return;
 
 #ifdef EMMALLOC_VERBOSE
-  MAIN_THREAD_ASYNC_EM_ASM(console.log('free(ptr=0x'+($0>>>0).toString(16)+')'), ptr);
+  _emscripten_outf("free(ptr=%p)", ptr);
 #endif
 
-  uint8_t *regionStartPtr = (uint8_t*)ptr - sizeof(uint32_t);
+  uint8_t *regionStartPtr = (uint8_t*)ptr - sizeof(size_t);
   Region *region = (Region*)(regionStartPtr);
-  assert(HAS_ALIGNMENT(region, sizeof(uint32_t)));
+  assert(HAS_ALIGNMENT(region, sizeof(size_t)));
 
   MALLOC_ACQUIRE();
 
-  uint32_t size = region->size;
+  size_t size = region->size;
 #ifdef EMMALLOC_VERBOSE
   if (size < sizeof(Region) || !region_is_in_use(region))
   {
     if (debug_region_is_consistent(region))
-      // LLVM wasm backend bug: cannot use MAIN_THREAD_ASYNC_EM_ASM() here, that generates internal compiler error
-      // Reproducible by running e.g. other.test_alloc_3GB
-      EM_ASM(console.error('Double free at region ptr 0x' + ($0>>>0).toString(16) + ', region->size: 0x' + ($1>>>0).toString(16) + ', region->sizeAtCeiling: 0x' + ($2>>>0).toString(16) + ')'), region, size, region_ceiling_size(region));
+      _emscripten_errf("Double free at region ptr 0x%p, region->size: %zx, region->sizeAtCeiling: %zx", region, size, region_ceiling_size(region));
```

Comment:
Oh I see, in wasm64 mode this LLVM bug would be hit on almost all prints? Does EM_JS() have that issue?

Reply:
I'll try to reproduce the bug and see if its still there..   I didn't run into that specific bug, but the more general issues with make EM_ASM blocks that work with both 64-bit and 32-bit wasm.

I'll try reverting all the formatting EM_ASM changes in this PR and try to find anther way to make the macros work in both modes.